src/rsh/mod.rs

```diff
@@ -59,51 +59,147 @@ pub fn run(initial_state: State) {
     }
 }
 
+#[derive(PartialEq)]
+enum BuildType {
+    None,
+    Single,
+    Double,
+}
+
+struct ParseResult {
+    result: Vec<String>,
+    completed: bool,
+    build_string: String,
+    build_type: BuildType,
+}
+
 fn parse_args(args: &String) -> Vec<String> {
-    let mut result: Vec<String> = Vec::new();
+    let mut parse_result: ParseResult = ParseResult {
+        result: Vec::new(),
+        completed: false,
+        build_string: String::from(""),
+        build_type: BuildType::None,
+    };
 
     if args.len() == 0 {
-        return result
-    }
-
-    let mut building_string: bool = false;
-    let mut build_string: String = String::from("");
-    for string in args.split_whitespace() {
-        if string.starts_with("\"") {
-            // the string is surrounded by quotes - "word"
-            if string.ends_with("\"") {
-                build_string.push_str(string);
-                result.push(build_string);
-
-                building_string = false;
-                build_string = String::from("");
-            // the string only begins with quote - "word
-            } else {
-                building_string = true;
-
-                build_string.push_str(string);
-                build_string.push(' ');
+        return parse_result.result;
+    }
+
+    let mut argstr = args.clone();
+    loop {
+        parse_string_into_vec(&argstr, &mut parse_result);
+
+        if (parse_result.completed) {
+            break;
+        }
+
+        match parse_result.build_type {
+            BuildType::Single => print!("quote> "),
+            BuildType::Double => print!("dquote> "),
+            _ => {}
+        }
+        io::stdout().flush();
+        argstr = String::from("");
+        io::stdin().read_line(&mut argstr).unwrap();
+    }
+
+    let result: Vec<String> = parse_result.result
+        .into_iter()
+        .filter(|s| s.len() > 0)
+        .collect();
+
+    result
+}
+
+fn parse_string_into_vec(string: &String, parse_result: &mut ParseResult) {
+    let &mut ParseResult { ref mut build_string,
```

Comment:
It was part convenience to not reference `parse_result` in every instance of its use, and also because I was having issues when not referencing the fields directly. I'd need to undo this to remember what the issue was unfortunately :(